Limit non-initial <input type=checkbox switch> styling to appearance:auto

annevk opened the pull request:
#### ac733e6cb304cd07c2456b71f170d8758755aad6
<pre>
Limit non-initial &lt;input type=checkbox switch&gt; styling to appearance:auto
<a href="https://bugs.webkit.org/show_bug.cgi?id=267334">https://bugs.webkit.org/show_bug.cgi?id=267334</a>

Reviewed by Aditya Keerthi and Tim Nguyen.

Per discussion in <a href="https://github.com/whatwg/html/issues/4180">https://github.com/whatwg/html/issues/4180</a> we do not
want styles intended for appearance:auto to leak into appearance:none.

Ideally we pay more attention to this with all form controls going
forward, but that is a much larger undertaking.

* LayoutTests/fast/forms/switch/display-expected.txt: Added.
* LayoutTests/fast/forms/switch/display.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/input-checkbox-switch.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/input-checkbox-switch.tentative.window.js:
* LayoutTests/platform/ios/TestExpectations:

Enable WebKit-specific switch tests on iOS too, except for those that
currently do not work.

* Source/WebCore/css/htmlSwitchControl.css:
(@namespace &quot;<a href="http://www.w3.org/1999/xhtml&quot">http://www.w3.org/1999/xhtml&quot</a>;;):
(input[type=&quot;checkbox&quot;][switch]::thumb, input[type=&quot;checkbox&quot;][switch]::track):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustStyle):
(WebCore::RenderTheme::adjustSwitchStyleDisplay const):
(WebCore::RenderTheme::adjustSwitchStyle const):
(WebCore::RenderTheme::adjustSwitchThumbOrSwitchTrackStyle const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustSwitchStyle const):

No longer skip most of the iOS style adjusting logic for non-auto sizes.

Canonical link: <a href="https://commits.webkit.org/273078@main">https://commits.webkit.org/273078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/734ae204350df8c1fbf8b46b4b2b44d67eb81649

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33985 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12775 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36617 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35052 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15159 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29851 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34482 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10799 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30317 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9420 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9520 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37925 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30866 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30658 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35633 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9651 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7582 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33533 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11436 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/29959 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10217 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4393 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10462 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -27,6 +27,7 @@ fast/events/ios [ Pass ]
 fast/events/ios/ipad [ Skip ]
 fast/forms/ios [ Pass ]
 fast/forms/ios/ipad [ Skip ]
+fast/forms/switch/display.html [ Pass ]
```
Slightly surprising we don't unskip the entirely directory on iOS?

annevk replied:
Yeah, I forgot to enable these tests after the iOS implementation landed. I'm also not entirely sure if all are compatible, but let's try and see what fails.

annevk replied:
Seems like they run just fine.

annevk replied:
Oh wait. I'm not sure why I thought they'd run fine. Nothing is painted so most of these will fail.

annevk replied:
I suggest I resolve this by explicitly skipping each test that fails because of painting in the iOS TestExpectations file. Since that's a trivial change, I hope this can be reviewed without that, but let me know if you want me to do that first.

pxlcoder replied:
Not sure I understand what you mean by "nothing is painted so most of these will fail". Looking at the failing tests on iOS, isn't it just that they need to be rewritten to use `UIHelper`, rather than only using `eventSender`?

In any case, feel free to set up / modify test expectations as necessary in this PR. We can look into rewriting tests and actually getting test parity separately.

pxlcoder commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -721,14 +721,26 @@ static bool renderThemePaintSwitchTrack(OptionSet<ControlStyle::State>, const Re
 }
 #endif
 
+// FIXME: Except for outline and disabled styling, we should deduplicate this with the generic code.
```
Seems like we should at least do that in this PR for the `switch (style.display())` duplication? I don't imagine it will be much extra work.

nt1m commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -723,12 +723,14 @@ static bool renderThemePaintSwitchTrack(OptionSet<ControlStyle::State>, const Re
-    auto size = std::max(style.computedFontSize(), logicalSwitchHeight);
-    style.setLogicalWidth({ size * (logicalSwitchWidth / logicalSwitchHeight), LengthType::Fixed });
-    style.setLogicalHeight({ size, LengthType::Fixed });
+    adjustSwitchStyleDisplay(style);
```
Is it OK that we run the code after this line for `!style.width().isAuto() && !style.height().isAuto()` ? Seems harmless but just want to be sure.

annevk replied:
Yeah, not running it was a bug.

annevk committed:
Limit non-initial <input type=checkbox switch> styling to appearance:auto
https://bugs.webkit.org/show_bug.cgi?id=267334

Reviewed by Aditya Keerthi and Tim Nguyen.

Per discussion in https://github.com/whatwg/html/issues/4180 we do not
want styles intended for appearance:auto to leak into appearance:none.

Ideally we pay more attention to this with all form controls going
forward, but that is a much larger undertaking.

* LayoutTests/fast/forms/switch/display-expected.txt: Added.
* LayoutTests/fast/forms/switch/display.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/input-checkbox-switch.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/input-checkbox-switch.tentative.window.js:
* LayoutTests/platform/ios/TestExpectations:

Enable WebKit-specific switch tests on iOS too, except for those that
currently do not work.

* Source/WebCore/css/htmlSwitchControl.css:
(@namespace "http://www.w3.org/1999/xhtml";):
(input[type="checkbox"][switch]::thumb, input[type="checkbox"][switch]::track):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustStyle):
(WebCore::RenderTheme::adjustSwitchStyleDisplay const):
(WebCore::RenderTheme::adjustSwitchStyle const):
(WebCore::RenderTheme::adjustSwitchThumbOrSwitchTrackStyle const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustSwitchStyle const):

No longer skip most of the iOS style adjusting logic for non-auto sizes.

Canonical link: https://commits.webkit.org/273078@main
